Add an actor for modifying the multipath configuration files

bmarzins opened the pull request:
This actor changes the the multipath configuration files so that they
will work in RHEL8, with as close a configuration to the RHEL7 version
as possible.

The actor silently comments out all removed configuration options,
specifically "hw_str_match", "ignore_new_boot_devs",
"new_bindings_in_boot" and "unpriv_sgio", since these all are no longer
necessary in RHEL8. It changes the option "detect_path_checker" to
"detect_checker", since the name has changed, and it removes
"hardware_handler" and "pg_timeout" from the overrides section, since
these no longer exist and never did anything in RHEL7.

If a features option contains "queue_if_no_path" it is removed, since it
has been deprecated in the config, and adds a "no_path_retry queue" option
instead, if the "no_path_retry" option did not already exist. This has
not change on the effective configuration.

If a device section with "all_devs" exist, the configuration values are
added to an overrides sections, since this replaces "all_devs" device
configs in RHEL8. If an overrides section already exists in the RHEL7
config, its values have priority over the all_devs values.

The actor also changes the values of "retain_attached_hw_handler",
"detect_checker", "detect_prio", and "reassign_maps" to the RHEL8
defaults, because this is necessary for proper functioning, but warns
the user of this.

Finally, if the path checker is set in the defaults section to something
other than "tur", it is not obvious how to upgrade the configuration,
and so the actor inhibits it.

pirat89 commented:
e2e tests

bmarzins commented:
Is those my job? If so, is there any documentation on how to write them, or examples? According to [The Leapp "Tests for actors" docs](https://leapp.readthedocs.io/en/latest/test-actors.html), 

> The QA team for the OAMG projects maintains an internal testing framework facilitating e2e tests.

That's all the info on them that I can find, and that's not much to go on.

vinzenz commented:
@bmarzins hehe no
This triggered our ci to run e2e tests ;-)

bmarzins commented:
review please

vojtechsokol commented:
@bmarzins
Hi, the tests were failing due to issue in our CI so I rescheduled them and requested a review from @oamg/developers on your behalf (your request was dissmissed because of the failing tests).

Rezney commented:
@bmarzins  Hi Ben, thank you very much for the PR. Just tested it and it is failing when "device-mapper-multipath" package is installed but multipath is not configured (missing "/etc/multipath.conf"):

```python
File "/usr/share/leapp-repository/repositories/system_upgrade/el7toel8/actors/multipathconfread/libraries/library.py", line 85, in _parse_config
    contents = multipathutil.read_config(path)
  File "/usr/share/leapp-repository/repositories/system_upgrade/el7toel8/libraries/multipathutil.py", line 15, in read_config
    with open(path, 'r') as f:
IOError: [Errno 2] No such file or directory: '/etc/multipath.conf'
```

Which I believe can happen from time to time. Could you please take a look? Cheers...

bmarzins commented:
> @bmarzins Hi Ben, thank you very much for the PR. Just tested it and it is failing when "device-mapper-multipath" package is installed but multipath is not configured (missing "/etc/multipath.conf"):
>  
> Which I believe can happen from time to time. Could you please take a look? Cheers...

Oops. Fixed the exception handling in multipathutil.py for that case, and added a unit test for missing config file.

Rezney commented:
@bmarzins could you please ask somebody from your team to do a review from multipath POV?

Rezney commented:
@AloisMahdal do you think we can have some integration test for this?

bmarzins commented:
> @bmarzins could you please ask somebody from your team to do a review from multipath POV?

I'm the device-mapper-multipath maintainer, so I'm the person that people would ask to do a multipath POV review.  Do I still need to find another person on the Storage Team to review this?

bmarzins commented:
I'll clean up the typos when your finished with your review.

Rezney commented:
@bmarzins If I use rhel7 default multipath.conf from your PR, after upgrade I get correctly:
`#       reassign_maps "yes" # Commented out by Leapp`



but if I use "/usr/share/doc/device-mapper-multipath-0.4.9/multipath.conf", uncomment the first default section and add reassign_maps as follows:

```
defaults {
        user_friendly_names yes
        find_multipaths yes
        reassign_maps "yes"
}
```

after upgrade reassign_maps is not commented out.

I did not spot any significant difference apart from another commented defaults section. Am I doing anything wrong? Cheers...

Rezney commented:
> > @bmarzins could you please ask somebody from your team to do a review from multipath POV?
> 
> I'm the device-mapper-multipath maintainer, so I'm the person that people would ask to do a multipath POV review. Do I still need to find another person on the Storage Team to review this?

I think it will not be needed then :)

bmarzins commented:
> @bmarzins If I use rhel7 default multipath.conf from your PR, after upgrade I get correctly:
> `# reassign_maps "yes" # Commented out by Leapp`
> 
> but if I use "/usr/share/doc/device-mapper-multipath-0.4.9/multipath.conf", uncomment the first default section and add reassign_maps as follows:
> 
> ```
> defaults {
>         user_friendly_names yes
>         find_multipaths yes
>         reassign_maps "yes"
> }
> ```
> 
> after upgrade reassign_maps is not commented out.
> 
> I did not spot any significant difference apart from another commented defaults section. Am I doing anything wrong? Cheers...

Nope. I did. It appears that while I was fighting with pylint to accept _nothing_to_do() from actors/multipathconfupdate/libraries/library.py, I ended up turning a chunk of that code into garbage. I'll fix that, and add some tests to verify that it's working.

bmarzins commented:
The _nothing_to_do() issues and the typos should be fixed now, and there are tests to verify that MultipathConfUpdate updates various configs that just have a single problem.

vinzenz commented on repos/system_upgrade/el7toel8/actors/multipathconfupdate/libraries/library.py:
```diff
@@ -0,0 +1,249 @@
+            config.overrides_hwhandler_exists or \
+            config.overrides_pg_timeout_exists or \
+            config.queue_if_no_path_exists or \
+            config.all_devs_section_exists:
```
About the too many boolean expressions:
```suggestion
    if any((config.hw_str_match_exists,
            config.ignore_new_boot_devs_exists,
            config.new_bindings_in_boot_exists,
            config.unpriv_sgio_exists,
            config.detect_path_checker_exists,
            config.overrides_hwhandler_exists,
            config.overrides_pg_timeout_exists,
            config.queue_if_no_path_exists,
            config.all_devs_section_exists)):
```

vinzenz commented on repos/system_upgrade/el7toel8/actors/multipathconfupdate/libraries/library.py:
```diff
@@ -0,0 +1,249 @@
+    if config.default_retain_hwhandler is False or \
+            config.default_detect_prio is False or \
+            config.default_detect_checker is False or \
+            config.reassign_maps is True:
```
About the too many boolean expressions:
```suggestion
    if not all((config.default_retain_hwhandler,
                config.default_detect_prio,
                config.default_detect_checker)) or config.reassign_maps:
```

vinzenz commented on repos/system_upgrade/el7toel8/actors/multipathconfupdate/libraries/library.py:
```diff
@@ -0,0 +1,249 @@
+        return False
+    if config.all_devs_options != []:
+        return False
+    return True
```
This could be doing about the same thing without the need for the linter instructions. Well it's up to you what you want to use :-)
```suggestion
def _nothing_to_do(config):
    result = True
    if config.default_path_checker and config.default_path_checker != "tur":
        result = False

    config_checks = ((
        (config.default_retain_hwhandler, False),
        (config.default_detect_prio, False),
        (config.default_detect_checker, False),
        (config.reassign_maps, True)
    ), (
        (config.hw_str_match_exists, True),
        (config.ignore_new_boot_devs_exists, True),
        (config.new_bindings_in_boot_exists, True),
        (config.unpriv_sgio_exists, True),
        (config.detect_path_checker_exists, True),
        (config.overrides_hwhandler_exists, True),
        (config.overrides_pg_timeout_exists, True),
        (config.queue_if_no_path_exists, True),
        (config.all_devs_section_ex, True)
    ))
    for checks in config_checks:
        if not result:
            break
        for option, value in checks:
            if option is value:
                result = False
                break
    if result and config.all_devs_options != []:
        result = False
    return result
```

bmarzins replied:
sure. There really isn't any reason to split config_checks into two tuples, with nested for loops (unless I'm missing something) but I'll do something like this.

vinzenz commented on repos/system_upgrade/el7toel8/actors/multipathconfupdate/libraries/library.py:
```diff
@@ -0,0 +1,249 @@
+        if line is ovr_line:
+            start = i + 1
+            break
+    if start is None:
```
```suggestion
    if start is None:
```
```suggestion
    if not start:
```

bmarzins replied:
sure

vinzenz commented on repos/system_upgrade/el7toel8/actors/multipathconfupdate/libraries/library.py:
```diff
@@ -0,0 +1,249 @@
+
+
+def _modify_features_line(line, value):
+    items = value.split()
```
Isn't this a duplication of the qinp removal in the other actor?

bmarzins replied:
The difference is that the MultipathCornfRead actor is getting info that will be used to create a while new section, so it doesn't want to keep the existing line data, like indentation or comments.  While in MultipathConfUpdate, it is modifying a line in-place, and wants to keep all that information.

vinzenz commented on repos/system_upgrade/el7toel8/models/multipathconffacts.py:
```diff
@@ -0,0 +1,58 @@
+       an overrides section'''
+
+
+class MultipathConfFacts(Model):
```
```suggestion
class MultipathConfFacts(Model):
    '''Model representing information from multipath configuration files'''
```

bmarzins replied:
sure

bmarzins commented on repos/system_upgrade/el7toel8/actors/multipathconfcheck/libraries/library.py:
```diff
@@ -0,0 +1,130 @@
+    options = {"default_path_checker": None, "detect_prio": None,
+               "detect_path_checker": None, "reassign_maps": None,
+               "retain_attached_hw_handler": None}
+    for config in configs:
```
The way multipath's configuration works, it processes the main file /etc/multipath.conf first, and then all the files in the config dir in alphabetical order, and each file can override options in files that came before. If you look at _parse_config_dir in the MultipathConfRead library, it sorts the files, and then reads them, so that they are added to the configs list in order.

bmarzins commented on repos/system_upgrade/el7toel8/actors/multipathconfcheck/libraries/library.py:
```diff
@@ -0,0 +1,130 @@
+    for i in range(1, len(paths) - 1):
+        path_str += ", " + paths[i]
+    path_str += " and " + paths[len(paths) - 1]
+    return path_str
```
neat. sure.

bmarzins commented on repos/system_upgrade/el7toel8/actors/multipathconfread/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+
+def _parse_config(path):
+    contents = multipathutil.read_config(path)
+    if contents is None:
```
sure

bmarzins replied:
Actually, while it doesn't make much difference in the end, I think that the reader should treat an empty config file different than a file it failed to read, so I'm keeping this.

vinzenz replied:
ok with me

bmarzins commented on repos/system_upgrade/el7toel8/actors/multipathconfread/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+                section = None
+            continue
+        if data.type == data.TYPE_SECTION_START:
+            if section is None:
```
sure

bmarzins commented on repos/system_upgrade/el7toel8/actors/multipathconfread/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+                    _add_options(conf.all_devs_options, device_options)
+                in_all_devs = False
+                device_options = []
+            elif section is not None:
```
sure

bmarzins commented on repos/system_upgrade/el7toel8/actors/multipathconfread/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+            if not path.endswith('.conf'):
+                continue
+            conf = _parse_config(path)
+            if conf is not None:
```
sure

bmarzins commented on repos/system_upgrade/el7toel8/actors/multipathconfread/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+def get_multipath_conf_facts(config_file="/etc/multipath.conf"):
+    res_configs = []
+    conf = _parse_config(config_file)
+    if conf is None:
```
sure

bmarzins commented on repos/system_upgrade/el7toel8/actors/multipathconfread/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+
+
+def _options_match(overrides_opt, all_devs_opt):
+    if overrides_opt == "detect_path_checker" and \
```
sure
